### PR TITLE
Snap 277 - Add maximum size option in S2TBX cache

### PR DESCRIPTION
**Pull request by Florian1209**
The user can active a new option : the cache size checking over the time (checked every minutes).
Then the maximum cache size can be adjusted. In case of exceeding cache size, the oldest product files should be deleted until the cache size has been reached 75% of the specified maximum.

The function is embedded in thread and the user help has been updated.

### CODE REVIEW

File: s2tbx-cache-ui/src/main/java/org/esa/s2tbx/dataio/s2/preferences/ui/cache/S2CacheOptionsPanel.java [hunk 1]
```diff
@@ -63,6 +69,23 @@ private void initComponents() {
         comboPanel.add(box);
         comboPanel.add(tableLayout.createHorizontalSpacer());
         comboPanel.add(informationLabel);
+        limitedSizeCacheField = new DoubleTextField("50");
```
Comment (marpet) [inline s2tbx-cache-ui/src/main/java/org/esa/s2tbx/dataio/s2/preferences/ui/cache/S2CacheOptionsPanel.java:76]: Use S2CacheUtils.SENTINEL_2_CACHE_MAX_SIZE_DEFAULT

**Review by marpet — APPROVED** [review]

File: s2tbx-cache-ui/src/main/java/org/esa/s2tbx/dataio/s2/preferences/ui/cache/S2CacheOptionsPanel.java (hunk 1 above)
Comment (Florian1209) [reply s2tbx-cache-ui/src/main/java/org/esa/s2tbx/dataio/s2/preferences/ui/cache/S2CacheOptionsPanel.java]: I update.
Thanks!

File: s2tbx-cache/src/main/java/org/esa/s2tbx/dataio/cache/S2CacheSizeChecking.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * @param period The period between two size checking in minute
+     */
+    public synchronized void launchCacheSizeChecking(double releasedSpacePurcent, int period) {
+        final ThreadExecutor executor = new ThreadExecutor();
+            try {
+                ThreadRunnable runnable = new ThreadRunnable() {
+                    @Override
+                    public void process() throws Exception {
+                        double byteToGigaByte = 0.000000001;
+                        while(!Thread.currentThread().isInterrupted()){
```
Comment (lnicola) [inline s2tbx-cache/src/main/java/org/esa/s2tbx/dataio/cache/S2CacheSizeChecking.java:46]: This is practically an infinite loop, `Thread.isInterrupted()` probably isn't that useful here. The code could use a condition variable to wait for exit notification, but setting up a timer would be better (and wouldn't hold up a thread pool thread indefinitely).

In addition, it looks like this _could_ presumably remove files which aren't in use, which might not be great.

File: s2tbx-cache/src/main/java/org/esa/s2tbx/dataio/cache/S2CacheSizeChecking.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    public synchronized void setParameters(boolean checkingEnable,double limitSizeCache ) {
+        this.checkingEnable = checkingEnable;
+        this.limitSizeCache = limitSizeCache;
+    }
+
+    /**
+     * Method check the S2cache size over time. The method is executed in thread.
+     * Each period time, the cache size is checked.
+     * The x% of the oldest files should be deleted in case of cache oversize.
+     * @param releasedSpacePurcent The space purcentage of the oldest files should be deleted
```
Comment (lnicola) [inline s2tbx-cache/src/main/java/org/esa/s2tbx/dataio/cache/S2CacheSizeChecking.java:36]: `releasedSpacePercent`

File: s2tbx-cache/src/main/java/org/esa/s2tbx/dataio/cache/S2CacheSizeChecking.java (hunk 2 above)
Comment (marpet) [reply s2tbx-cache/src/main/java/org/esa/s2tbx/dataio/cache/S2CacheSizeChecking.java:46]: Your suggestions might work, but actually this is started from S2CacheActivator.start() and should be stopped by S2CacheActivator.stop()

Comment (lnicola) [reply s2tbx-cache/src/main/java/org/esa/s2tbx/dataio/cache/S2CacheSizeChecking.java:46]: I suppose the `S2CacheActivator` might just as well cancel the timer.

**Pull request merged**